README.md
# PortfolioAllocation ([Changelog](changelog.md))

[![npm](https://img.shields.io/npm/v/portfolio-allocation.svg)](https://www.npmjs.com/package/portfolio-allocation)
[![Travis Build Status](https://travis-ci.org/lequant40/portfolio_allocation_js.svg?style=flat)](https://travis-ci.org/lequant40/portfolio_allocation_js)
[![Feature Requests](https://feathub.com/lequant40/portfolio_allocation_js?format=svg)](https://feathub.com/lequant40/portfolio_allocation_js)

PortfolioAllocation is a JavaScript library designed to help constructing financial portfolios made of several assets: bonds, commodities, cryptocurrencies, currencies, exchange traded funds (ETFs), mutual funds, stocks...

When constructing such portfolios, one of the main problems faced is to determine the proportion of the different assets to hold.

PortfolioAllocation solves this problem using mathematical optimization algorithms.

Do not hesitate to [report any bug](https://github.com/lequant40/portfolio_allocation_js/issues) / [request additional features](https://feathub.com/lequant40/portfolio_allocation_js) !

## Features

- Compatible with Google Sheets
- Compatible with any browser supporting ECMAScript 5 for front-end development
- Compatible with [Node.js](https://nodejs.org/) for back-end development
- Code continuously tested and integrated by [Travis CI](https://travis-ci.org/)
- Code documented for developers using [JSDoc](http://usejsdoc.org/)


## Usage

### Usage in Google Sheets

***Note: Examples of how to integrate PortfolioAllocation in Google Sheets are provided in [this spreadsheet](https://docs.google.com/spreadsheets/d/1ScrwSjr9EgwXfRyPN4IaqVxZvDnqw-hWvVQcJ9Ak590).***

First, make the PortfolioAllocation functions available in your spreadsheet script:
- *(Recommended)* Import PortfolioAllocation as any other [external Google Apps Script library](https://developers.google.com/apps-script/guide_libraries), using the Script ID ***1cTTAt3VZRZKptyhXtjF3EK-jKagdTUl6t29Pc7YidrC5m5ABR6LUy8sC***
- *(Alternative)* Copy/paste the JavaScript files from the [dist/gs directory](https://github.com/lequant40/portfolio_allocation_js/tree/master/dist/gs)

Then, you can call these functions your preferred way in your spreadsheet script.

Here is an example through a wrapper function, to which spreadsheet data ranges (e.g. A1:B3) can be provided directly:
```js
function computeERCPortfolioWeights(covarianceMatrix) {
  // Note: The input range coming from the spreadsheet is directly usable.
    
  // Compute the ERC portfolio weights
  var ercWeights = PortfolioAllocation.equalRiskContributionWeights(covarianceMatrix);
  
  // Return them to the spreadsheet
  return ercWeights;
}
```

### Usage in a browser

***Note: PortfolioAllocation is delivered through the CDN [jsDelivr](https://www.jsdelivr.com/), at [this URL](https://cdn.jsdelivr.net/npm/portfolio-allocation/dist/portfolio_allocation.dist.min.js).***

Include PortfolioAllocation minified source file in an HTML page, and you are done:
```html
<script src="https://cdn.jsdelivr.net/npm/portfolio-allocation/dist/portfolio_allocation.dist.min.js" type="text/javascript"></script>
<script type="text/javascript">
  var w = PortfolioAllocation.riskBudgetingWeights([[0.1,0], [0,0.2]], [0.25, 0.75]);
</script>
```

### Usage with Node.js

***Note: PortfolioAllocation is delivered as the [npm](https://www.npmjs.com/) package [portfolio-allocation](https://www.npmjs.com/package/portfolio-allocation).***

First, [declare PortfolioAllocation as a dependency in your project's `package.json` file](https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file), using the package name ***portfolio-allocation***.

Then, this is standard Node.js:

```js
var PortfolioAllocation = require('portfolio-allocation');
var w = PortfolioAllocation.riskBudgetingWeights([[0.1,0], [0,0.2]], [0.25, 0.75]);
```


## Included algorithms

### Portfolio allocation and optimization algorithms

- Equal weights (EW)  
  Analyzed by Victor DeMiguel and al. in their research paper [Optimal Versus Naive Diversification: How Inefficient is the 1/N Portfolio Strategy?](https://doi.org/10.1093/rfs/hhm075).

- Inverse volatility (IV)  
  Described by Raul Leote de Carvalho and al. in the research paper [Demystifying Equity Risk-Based Strategies: A Simple Alpha Plus Beta Description](https://doi.org/10.3905/jpm.2012.38.3.056).

- Equal risk contributions (ERC) and risk budgeting (RB)  
  Extensively studied by [Thierry Roncalli](http://www.thierry-roncalli.com/) and al in misc. research papers ([The properties of equally weighted risk contribution portfolios](https://doi.org/10.3905/jpm.2010.36.4.060), [Managing Risk Exposures Using the Risk Budgeting Approach](https://ssrn.com/abstract=2009778), [Constrained Risk Budgeting Portfolios](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3331184)...).

- Equal risk bounding (ERB)  
  Described in the research paper [Equal Risk Bounding is better than Risk Parity for portfolio selection](https://doi.org/10.1007/s10898-016-0477-6) by Francesco Cesarone and Fabio Tardella, the ERB portfolio is an ERC portfolio possibly not containing all the assets in the considered universe.

- Cluster risk parity (CRP)  
  Discovered by [David Varadi](https://cssanalytics.wordpress.com/) and [Michael Kapler](http://systematicinvestor.wordpress.com/), the CRP portfolio combines the usage of a clustering algorithm (for instance, the Fast Threshold Clustering Algorithm - FTCA - of David Varadi) with the ERC portfolio.

- Most diversified portfolio (MDP)  
  Introduced in the research paper [Toward Maximum Diversification](https://doi.org/10.3905/JPM.2008.35.1.40) by [Yves Choueifaty](http://www.tobam.fr/yves-choueifaty/) and al., it maximizes what the authors call the diversification ratio, which is the weighted average of the assets volatilities divided by the portfolio total volatility.

- Minimum correlation algorithm (MCA)  
  Discovered by [David Varadi](https://cssanalytics.wordpress.com/), the MCA portfolio is meant to be an approximation of the MDP portfolio.

- Mean-variance optimization (MVO)  
  Based on the modern portfolio theory described by Harry M. Markowitz in numerous articles and books ([Portfolio Selection: Efficient Diversification of Investments](https://www.jstor.org/stable/j.ctt1bh4c8h)...), the portfolio obtained through mean-variance optimization is mean-variance efficient, that is, for a given level of return, it possesses the lowest attainable volatility and for a given level of volatility, it possesses the highest attainable return.

- Global minimum variance (GMV)  
  The leftmost portfolio on the mean-variance efficient frontier, the GMV portfolio possesses the smallest attainable volatility among all the mean-variance efficient portfolios.

- Proportional minimum variance algorithm (MVA)  
  Discovered by [David Varadi](https://cssanalytics.wordpress.com/), the MVA portfolio is meant to be an approximation of the GMV portfolio.

- Minimax portfolio  
  Introduced by Martin Young in the research paper [A Minimax Portfolio Selection Rule with Linear Programming Solution](http://www.jstor.org/stable/2634472), the minimax portfolio uses the minimum return as a measure of risk instead of the variance as in the Markowitz framework.

- Random portfolio  
  Random portfolios are generally used to benchmark the performances of portfolio allocation and optimization algorithms, as pioneered by Ronald J. Surz in the article [Portfolio Opportunity Distributions](https://doi.org/10.3905/joi.3.2.36) and latter complemented by Patrick Burns in the article [Random Portfolios for Performance Measurement](https://doi.org/10.1007/3-540-36626-1_11).

- Maximum Sharpe ratio (MSR), a.k.a. (Lintner) tangency portfolio  
  Introduced by John Lintner in the research paper [The Valuation of Risk Assets and the Selection of Risky Investments in Stock Portfolios and Capital Budgets](https://www.jstor.org/stable/1924119), the MSR portfolio possesses the highest Sharpe ratio among all the mean-variance efficient portfolios.

- Random subspace mean-variance optimization (RSO-MVO)  
  Discovered by [David Varadi](https://cssanalytics.wordpress.com/) and formally studied by Benjamin J. Gillen in the research paper [Subset Optimization for Asset Allocation](https://authors.library.caltech.edu/79336/), the RSO-MVO portfolio combines the usage of a [random subspace method](https://en.wikipedia.org/wiki/Random_subspace_method) with a mean-variance optimization method.

- Minimum tracking error portfolio, a.k.a. index tracking portfolio  
  The index tracking portfolio aims at replicating the performances of a given stock market index, or more generally of a given benchmark, with a limited number of its constituents thanks to cardinality constraints.
  
- Best constantly rebalanced portfolio (BCRP)  
  The BCRP portfolio is a portfolio determined in hindsight to benchmark the performances of online portfolio selection algorithms.

  
### Misc. other algorithms
- Post-processing of numerical portfolio weights   
  The weights obtained through a portfolio optimization algorithm (e.g. w = 0.123456789) need in practice to be either rounded off (e.g. w = 0.12) or converted into an integer number of shares (e.g. q = 10 shares, or q = 2 lots of 100 shares).

- Computation of the mean-variance efficient frontier   
  The continuous set of all mean-variance efficient portfolios (the mean-variance efficient frontier) as well as its generating discrete set (the set of corner portfolios) can both be efficiently computed thanks to a specialized algorithm developed by Harry M. Markowitz: [the critical line method](https://web.stanford.edu/~wfsharpe/mia/opt/mia_opt3.htm).

- Computation of the nearest portfolio on the mean-variance efficient frontier    
  Thanks to the Markowitz's critical line method, it is possible to exactly compute the nearest mean-variance efficient portfolio of any given portfolio.

- Generation of perturbed mean vectors, variances and correlation matrices   
  As demonstrated in [The effect of errors in means, variances, and covariances on optimal portfolio choice](https://jpm.pm-research.com/content/19/2/6) by Vijay Chopra and William Ziemba, the impact of estimation errors in expected returns, variances and correlation matrices on the output of a portfolio optimization algorithm can be significant, so that it is useful to have an algorithm to perturb these quantities for portfolio weights sensitivity analysis.

- Generic random subspace optimization   
  A direct extension of the RSO-MVO method, allowing to use the random subspace optimization method with any portfolio optimization method.

- Generic numerical optimization   
  When no specialized algorithm exist to solve a particular portfolio optimization problem, it is always possible to use a generic numerical optimization algorithm (e.g., grid search on the simplex).

- Random generation of mean vectors, variances and correlation matrices   
  When implementing portfolio optimization algorithms, the capability to generate random mean vectors, random variances and random correlation matrices can be of great help.

- Computation of shrinkage estimators for mean vectors and covariance matrices   
  Shrinkage estimators for mean vectors can help to reduce the estimation errors in expected returns as observed in DeMiguel et al. [Size Matters: Optimal Calibration of Shrinkage Estimators for Portfolio Selection](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1891847), and shrinkage estimators for covariance matrices like Ledoit-Wolf's [A Well-Conditioned Estimator for Large-Dimensional Covariance Matrices](http://www.ledoit.net/ole1_abstract.htm) provide an elegant solution to the problem of the ill-conditioning and non-invertibility of sample covariance matrices.

- Repairing indefinite or positive semi-definite correlation matrices   
  With certain use-cases (e.g. alteration of correlation pairs for stress testing), correlation matrices might lose their positive (semi) definiteness, which is possible to recover thanks for instance to the spectral method described by Rebonato and Jaeckel in their paper [The Most General Methodology to Create a Valid Correlation Matrix for Risk Management and Option Pricing Purposes](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1969689) or to other methods like the computation of the nearest correlation matrix addressed by Nicholas J. Higham in his paper [Computing the nearest correlation matrix - A problem from Finance](https://www.maths.manchester.ac.uk/~higham/narep/narep369.pdf).

  
## Documentation

The code is heavily documented, using [JSDoc](http://usejsdoc.org/).

That being said, the documentation is rather for developer's usage, so that in case of trouble to use any algorithm, do not hesitate to ask for support !


## Contributing

### Fork the project from [Github](https://github.com/)...

### Install the [Grunt](http://gruntjs.com/) dependencies and command line

```
npm install
npm install -g grunt-cli
```

### Develop...

### Compile and test

- The following commands generates the files to be used inside a browser or with Node.js in the `dist` directory:

```
grunt deliver-dev
grunt deliver-dist
```

- The following command generates the files to be used in Google Sheets in the `dist\gs` directory:

```
grunt deliver-gs
```


### Submit a pull-request...


## License

[MIT License](https://en.wikipedia.org/wiki/MIT_License)



changelog.md
### 0.0.12 - xx/yy/2020

### 0.0.11 - 09/10/2020
- Cosmetic updates related correlation matrices fixing, and added a unit test to cover the infinite precision of the nearest correlation matrix computation
- Improved the numerical stability of the Constrained Risk Budgeting algorithm: added auto-expanding bisection interval, and early stopping criterion
- Added methods to swap two rows or two columns of a matrix
- Added an LU decomposition method for square matrices, with full pivoting
- Added a method to solve lower triangular linear systems
- Reworked the internals of the critical line algorithm, to solve the KKT linear systems with a LU decomposition with full pivoting instead, to guard against numerical issues
- Reworked the internals of the critical line algorithm to allow for equal returns; the only case in which the CLA algorithm will not work is if the covariance matrix is not invertible, if there are more than 1 asset IN in the maximum return portfolio, and if the associated KKT matrix is also non invertible
- Fixed a bug in the GSMO optimization method, which could seemingly lead to cycling (or at least required an unreasonably huge number of iterations before convergence)
- Added the critical line method as additional optimization method for the GMV portfolio
- Added the critical line method as additional optimization method for the MDP portfolio
- Improved the numerical precision of the corner portfolios, of the mean-variance efficient portfolios, and of the maximum Sharpe ratio portfolios
- Added a RSO-MSR optimization method
- Redesigned the asset allocation non-regression to factorize the tests on portfolio weights using an equality function
- Redesigned the internals of the mean-variance optimization method, to automatically selects the critical line algorithm, or the GSMO algorithm, as optimization routine; this impacts all portfolio computations using MVO
- Allowed for an infinite number of cycles in the risk budgeting computation
- Added lower bounds and upper bounds constraints in the equal risk bounding method
- Updated the covariance matrix computation method to be able to use Bessel's correction with Ledoit-Wolf shrinkage (= non-backward compatible change)
- Updated the mean vector computation method to be consistent with the covariance computation method w.r.t. the regularizationMethod parameter (= non-backward compatible change)
- Updated the returns computation method to generate an error in case of division by zero
- Added a method to test that a matrix is a covariance matrix
- Fixed a bug in Cholesky decomposition, which wrongly allowed for semi-definite positive matrices to be decomposed
- Added a method to compute a Cholesky decomposition for a positive semi-definite matrix
- Misc. checks added in the covariance matrix creation methods
- Improved the method to check a correlation matrix by allowing details on what property the matrix is missing
- Added a method to compute a covariance matrix from a supposed covariance matrix
- Updated the Threshold Accepting method to enable optimization in computations of successive values of the function f, and used this optimization with the postOptimizationWeights and minimumTrackingErrorWeights methods
- Removed partial investment constraints on MVO, Max Sharpe portfolios
- Improved the performances of the minimum tracking error portfolio in case of cardinality constraints 
- Removed the post optimization method "rationalizing" (= non backward compatible change), as this one is useless in practice
- Misc. performances and randomization improvements in the post optimization method internals
- Automatic computation of the portfolio target cash position, as well as management of levered portfolio, in the post optimization weights method
- Fixed a numerical bug in correlation matrix extraction from covariance matrix, which was not strictly symmetric and unit diagonal
- Polished the output of the nearest correlation matrix computation to have a strictly symmetric and unit diagonal matrix in output whatever the requested precision
- Misc. optimizations in the critical-line algorithm when assets returns are identical
- Misc. optimizations in the computation of efficient portfolios in case the GSMO algorithm is used
- Added the possibility to provide an initial point to the GSMO algorithm for warm start
- Fixed a bug in the projection on the efficient frontier in case the GSMO algorithm is used, which lead to the same efficient portfolios being computed on the efficient frontier, thus biaising the projection
- Removed partial investment constraint support from all MV-related algorithms (= non backward compatible changes), and introduced a proper portfolio exposure management in the RSO-MV related algorithms
- Removed partial investment constraint support from minimum tracking error algorithm, in case of no cardinality constraints (= non backward compatible changes)
- Fixed a bug in all MV-related algorithms where in case of tight min/max weight constraints, the algorithms would loop infinitely

### 0.0.10 - 14/09/2020
- Added the Jacobi method in order to compute eigenvalues and eigenvectors of real symmetric matrices
- Updated the method to generate random correlation matrices to output truly symmetric matrices, and not numerically symmetric matrices only
- Added a method to repair the (semi) definite-positiveness of correlation matrices, using spectral decomposition
- Change of method name for Ledoit-Wolf covariance matrix shrinkage to "linear-shrinkage" (= non-backward compatible change)
- Change of method name for DeMiguel mean vector shrinkage to "linear-shrinkage" (= non-backward compatible change)
- Added a method to repair the (semi) definite-positiveness of correlation matrices, using shrinkage towards a target matrix
- Added the possibility to provide standard deviations for random mean vectors and random variances vectors creation
- Added a method to randomly perturb a correlation matrix

lib/allocation/rb.js
/**
 * @file Functions related to risk budgeting portfolio.
 * @author Roman Rubsamen <roman.rubsamen@gmail.com>
 */

 
/* Start Wrapper private methods - Unit tests usage only */
/* End Wrapper private methods - Unit tests usage only */


/**
* @function riskBudgetingWeights
*
* @summary Compute the weights of the risk budgeting portfolio.
*
* @description This function returns the weights w_1,...,w_n associated to the fully invested and long-only portfolio
* of n assets with risk budgeting contraints, defined as the weights with the property that the contribution 
* of each asset to the risk of the portfolio is equal to a pre-determined budget weight.
*
* This portfolio is unique, provided the covariance matrix of the assets is definite positive.
* 
* The algorithm used is a cyclical coordinate descent, c.f. the second reference, whose convergence is guaranteed
* if the covariance matrix of the assets is definite positive.
*
* @see <a href="https://ssrn.com/abstract=2009778">Bruder, Benjamin and Roncalli, Thierry, Managing Risk Exposures Using the Risk Budgeting Approach (January 20, 2012).</a>
* @see <a href="https://arxiv.org/abs/1311.4057">Théophile Griveau-Billion, Jean-Charles Richard, Thierry Roncalli; A Fast Algorithm for Computing High-dimensional Risk Parity Portfolios. eprint arXiv:1311.4057</a>
* 
* @param {Matrix_|Array.<Array.<number>>} sigma the covariance matrix (sigma_ij),i,j=1..n of the n assets in the considered universe, square Matrix or array of n array of n real numbers statisfying sigma[i-1][j-1] = sigma_ij.
* @param {Array.<number>} rb the risk budgets, array of n real strictly positive numbers summing to one.
* @param {object} opt the optional parameters for the algorithm.
* @param {number} opt.eps the tolerance parameter for the convergence of the algorithm, a strictly positive real number; defaults to 1e-8.
* @param {number} opt.maxIter the maximum number of iterations of the algorithm, a strictly positive natural integer; defaults to 10000.
* @return {Array.<number>} the weights corresponding to the risk budgeting portfolio, array of n real numbers.
*
* @example
* riskBudgetingWeights([[0.1,0], [0,0.2]], [0.25, 0.75], {eps: 1e-10, maxIter: 10000});
* // [~0.45, ~0.55]
*/
self.riskBudgetingWeights = function (sigma, rb, opt) {
	// Decode options
	if (opt === undefined) {
		opt = {};
	}
	var eps = opt.eps || 1e-8;
	var maxIterations = opt.maxIter || 10000;
	
	// Convert sigma to matrix format
	var sigma = new Matrix_(sigma);
	
	// Convert rb to vector format
	var rb = new Matrix_(rb);

	// TODO: Checks, if enabled
	// Check that diagonal entries of sigma are strictly positive
	// Check that sigma is symmetric and positive definite
	// Check that rb contains strictly positive numbers summing to one
	// Check that sigma and rb are rows compatible


	// ------
	var nbAssets = sigma.nbRows;
	
	// Initial point for the algorithm is an equal weight vector
	var x = Matrix_.ones(nbAssets, 1).normalize();
	
	// Preparational computations
	var sigma_x = Matrix_.product(sigma, x); // SIGMA*x
	var s_x = Math.sqrt(Matrix_.vectorDotProduct(sigma_x, x)); // sigma(x)
	
	// Main loop until convergence, guaranteed as per hypotheses on sigma and b
	var iter = 0;
	var converged = false;
	while (!converged) {
        // Convergence condition is false if any of the coordinate-wise convergence condition is false
    	converged = true;
    	
    	for (var i = 1; i <= nbAssets; ++i) {
    	    // Define the coefficients of the second order polynomial ax_i^2 + b_ix + c, c.f. the second reference
    	    var a = sigma.getValueAt(i,i); // sigma_i^2, always > 0
    	    var b = sigma_x.getValueAt(i,1) - x.getValueAt(i,1) * sigma.getValueAt(i,i); // (SIGMA*x)_i - x_i*sigma_i^2, might be any sign
    	    var c = -rb.getValueAt(i,1) * s_x; // -b_i * sigma(x), always < 0
    	    
    	    // Extract the strictly positive root x_i^* of the equation ax_i^2 + bx_i + c = 0, using a stable numerical formula
    	    var b_p = b/2; // reduced discriminant
			var sign_b_p = (b_p >= 0) ? 1 : -1; // Math.sign is not supported everywhere plus it is mandatory that for b_p == 0 this returns 1
			var disc = b_p*b_p - a*c;
			if (disc < 0) {
			    throw new Error('Discriminant not positive during iteration ' + iter + ', covariance matrix might not be definite positive');
			}
    	    var q = -(b_p + sign_b_p * Math.sqrt(disc));
    	    var r1 = q/a;
    	    var r2 = c/q;
    	    var xi_star = r1 > 0 ? r1 : r2;
    	    
    	    // Update the weights
    	    x.setValueAt(i,1,xi_star);
    	    
    	    // Compute the updated SIGMA*x and x'*SIGMA*x products for convergence condition evaluation + next loop evaluation
    	    sigma_x = Matrix_.product(sigma, x)
    	    s_x = Math.sqrt(Matrix_.vectorDotProduct(sigma_x, x));
    	    
    	    // Update the convergence condition: |RC_i* - b_i| <= eps, i = 1..nbAssets
    	    var rci_star = x.getValueAt(i,1) * sigma_x.getValueAt(i,1) / s_x;
    	    if (Math.abs(rci_star - rb.getValueAt(i,1)) > eps) {
    	        converged = false;
    	    }
        }
		
		// Update the number of iterations
		++iter;
		
		// Check the number of iterations
		if (iter >= maxIterations) {
			throw new Error('maximum number of iterations reached: ' + maxIterations);
		}
	}
	
	// Return the computed weights, after normalization
	x = x.normalize();
	return x.toArray();
}

